### PR TITLE
BXMSPROD-781 PipelineHelper.retry added

### PR DESCRIPTION
**Pull request by Ginxo**
https://issues.redhat.com/browse/BXMSPROD-781
depends on:

- https://github.com/kiegroup/jenkins-pipeline-shared-libraries/pull/33
- https://github.com/kiegroup/jenkins-pipeline-shared-libraries/pull/42

naginator plugin is not available for pipelines https://plugins.jenkins.io/naginator/

### CODE REVIEW

File: Jenkinsfile.nightly [hunk 1]
```diff
@@ -28,23 +27,27 @@ pipeline {
...
+                            println "Reading file ${repositoryListPath} and merging it with ${PRODUCTION_PROJECT_LIST} jenkins file"
+                            def communityProjectListFile = readFile repositoryListPath
+                            def projectCollection = communityProjectListFile.readLines()
+                            def productionProjectListFile = readFile "${env.PRODUCTION_PROJECT_LIST}"
+                            projectCollection.addAll(productionProjectListFile.readLines())
+                            projectCollection.removeAll { 'droolsjbpm-tools'.equals(it.toLowerCase()) } // The droolsjbpm-tools is skiped
+                            println "File ${repositoryListPath} and ${env.PRODUCTION_PROJECT_LIST} jenkins file merged in ${projectCollection}"
+                            nightlyStages(SETTINGS_XML_ID, projectCollection, "rhba/nightly", projectVariableMap)
+                        }
+                    }, 2, 720*60)
```
Comment (mareknovotny) [inline Jenkinsfile.nightly:50]: what is the reasoning behind this timeout 720 minutes aka 12 hours reattempt ?

**Comment by almope** [comment]
@Ginxo leaving here a question to keep it tracked alongside and try to persuade you to look for a different approach and to not rush with this one:

What is the exact issue we're experiencing that requires a custom feature to retry the complete job rebuild?

Sorry, it's not clear from the Jira issue.

File: Jenkinsfile.nightly (hunk 1 above)
Comment (Ginxo) [reply Jenkinsfile.nightly:50]: @mareknovotny 720 was the original value, according to the latest executions it's between 4 and 6,5 hours, so we can decrease it to 8 hours?

**Comment by Ginxo** [comment]
@almope the requirement was just to retry job in case of failure. It can be due to git problems or even openstack outage, so in order to assure as much as possible nightly will be performed we agreed to just retry it.
Maybe @mareknovotny can add something here ¿?

**Comment by mareknovotny** [comment]
yes that was that. Just one comment we should retry only in case these infra problem is that possible to check and have this logic there?

**Comment by mareknovotny** [comment]
@Ginxo the PR build https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/blue/organizations/jenkins/PROD%2Fnightly/detail/PR-1346/10/pipeline#log-33 says:
```
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:

WorkflowScript: 30: unable to resolve class PipelineHelper 

 @ line 30, column 42.

            def pipelineHelper = new Pipeli

                                 ^


1 error


	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:310)

	at org.codehaus.groovy.control.CompilationUnit.applyToSourceUnits(CompilationUnit.java:958)

	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:605)

	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:554)

	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:298)

	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:268)

	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:688)

	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)

	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.doParse(CpsGroovyShell.java:142)

	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.reparse(CpsGroovyShell.java:127)

	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.parseScript(CpsFlowExecution.java:561)

	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.start(CpsFlowExecution.java:522)

	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:327)

	at hudson.model.ResourceController.execute(ResourceController.java:97)

	at hudson.model.Executor.run(Executor.java:429)

Finished: FAILURE
```

**Comment by Ginxo** [comment]
@winklerm there is a chance to do this
```

        stage('Test Timeout') {
            steps {
                retry(count: 3) {
                    timeout(time: 6, unit: 'SECONDS') {
                        script {
                            try {
                                echo 'Sleeping'
                                sleep 7
                                echo 'Slept'
                            } catch (FlowInterruptedException e) {
                                error 'Timeout!'
                            }
                        }
                    }
                }
            }
        }
```
You can check result here https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/emingora/job/BXMSPROD-781/job/retry-test/43/console
so we have both things retry + timeout per shot in a more common way

**Comment by winklerm** [comment]
@Ginxo Nice, thank you for trying it out, looks good to me!

**Pull request merged**